lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -2739,6 +3003,154 @@ impl DeviceReadResponse {
+                device_io_request,
+                NTSTATUS::to_u32(&io_status).unwrap(),
+            ),
+            length,
+        }
+    }
+
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        w.extend_from_slice(&self.device_io_reply.encode()?);
+        w.write_u32::<LittleEndian>(self.length)?;
+        // 1 byte padding
+        w.write_u32::<LittleEndian>(0)?;
+        Ok(w)
+    }
+}
+
+/// 2.2.3.4.9 Client Drive Set Information Response (DR_DRIVE_SET_INFORMATION_RSP)
+/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/16b893d5-5d8b-49d1-8dcb-ee21e7612970
+#[derive(Debug)]
+#[allow(dead_code)]
+struct ClientDriveSetInformationResponse {
+    device_io_reply: DeviceIoResponse,
+    /// This field MUST be equal to the Length field in the Server Drive Set Information Request (section 2.2.3.3.9).
+    length: u32,
+}
+
+impl ClientDriveSetInformationResponse {
+    fn new(device_read_request: &ServerDriveSetInformationRequest, io_status: NTSTATUS) -> Self {
+        Self {
+            device_io_reply: DeviceIoResponse::new(
+                &device_read_request.device_io_request,
+                NTSTATUS::to_u32(&io_status).unwrap(),
+            ),
+            length: device_read_request.set_buffer.len() as u32,
+        }
+    }
+
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        w.extend_from_slice(&self.device_io_reply.encode()?);
+        w.write_u32::<LittleEndian>(self.length)?;
+        // 1 byte padding
+        w.write_u32::<LittleEndian>(0)?;
+        Ok(w)
+    }
+}
+
+/// 2.2.3.3.9 Server Drive Set Information Request (DR_DRIVE_SET_INFORMATION_REQ)
+/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/b5d3104b-0e42-4cf8-9059-e9fe86615e5c
+#[derive(Debug)]
+#[allow(dead_code)]
+struct ServerDriveSetInformationRequest {
+    /// The MajorFunction field in the DR_DEVICE_IOREQUEST header MUST be set to IRP_MJ_SET_INFORMATION.
+    device_io_request: DeviceIoRequest,
+    file_information_class_level: FileInformationClassLevel,
+    /// TODO(LKozlowski): change to enum?
+    set_buffer: Vec<u8>,
+}
+
+impl ServerDriveSetInformationRequest {
+    fn decode(device_io_request: DeviceIoRequest, payload: &mut Payload) -> RdpResult<Self> {
+        let file_information_class_level =
+            FileInformationClassLevel::from_u32(payload.read_u32::<LittleEndian>()?)
+                .ok_or_else(|| invalid_data_error("failed to read FileInformationClassLevel"))?;
+
+        let valid_levels = vec![
+            FileInformationClassLevel::FileBasicInformation,
+            FileInformationClassLevel::FileEndOfFileInformation,
+            FileInformationClassLevel::FileDispositionInformation,
+            FileInformationClassLevel::FileRenameInformation,
+            FileInformationClassLevel::FileAllocationInformation,
+        ];
+
+        if !valid_levels.contains(&file_information_class_level) {
```

Comment:
Maybe this is optimized away and wouldn't matter, but it feels like a match would be better than allocating memory for a vec and then iterating through it.